### ISSUE TITLE
Would you be willing to include lessons on capturing (closures / async blocks) and the `Fn*` traits?

### ISSUE THREAD

**Issue by junderw** [rust-lang/rustlings#1746]
One topic I see new people hit pretty early on is the capturing semantics and the `Fn*` traits.

I was thinking of adding a few lessons to show:

1. Capturing semantics without `move`
2. Capturing semantics with `move` (including ie. binding a reference to a variable before using it in the closure and a reference is moved)
3. How the `Fn*` traits are auto-implemented.
4. Briefly mention that async block capturing is similar to closure capturing, but that there is no `Fn*` traits for async-block-futures.

I wanted to read the room a bit before actually putting time into this. Do we think this might be a bit too advanced for rustlings? Would it depend on the actual content? Would anyone be willing to help review my lessons to try and get it over the finish line?

It's an issue I get asked about constantly, especially when dealing with GUI frameworks like egui that make use of lots of nested closures... it might help people intuit these issues more clearly.

Let me know what you think!

**Comment by divyaranjan1905** [comment]
> I think closures should be explained after traits and tests. This is in line with the Rust book as well.

Indeed, as I proposed in the issue, it would replace the currently 18th set (iterators), so would be after traits and tests.

**Comment by mo8it** [comment]
You are right, I forgot about this PR: https://github.com/rust-lang/rustlings/pull/1748

I agree @junderw, that PR only shows the basics. After talking about traits, I would like to have 3 additional exercises where you need to write your own functions that take a closure, one with `Fn`, one with `FnMut` and one with `FnOnce`. What do you think?

**Comment by divyaranjan1905** [comment]
> After talking about traits, I would like to have 3 additional exercises where you need to write your own functions that take a closure, one with `Fn`, one with `FnMut` and one with `FnOnce`. What do you think?

Yep, @mo8it I think that'll be in line with the chapter from The Book. I'm considering to keep the number of exercises to 4, or at most 5.

**Comment by rikettsie** [comment]
@divyaranjan1905, @junderw , when I wrote that couple of exercises last year (#1748), I considered them a gentle introduction
to closures.

I still think it's a good idea to have a taste of closures early in those chapters (_functions_ and _move_semantics_),
and definitely a very good thing to add a dedicated chapter later to build a deeper understanding (probably between _traits_ and _iterators_).

@mo8it, I updated the PR #1748 to compile to the new rustlings standard, if you think it's valuable to merge.

**Comment by junderw** [comment]
I rebased and added a closures2.rs to the PR in #2099
